bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/ColumnViewer.java

```diff
@@ -817,4 +841,194 @@ protected void setBusy(boolean busy) {
+	 * {@link IExpandableNode} we will return next block of limited elements to be
+	 * created.
+	 *
+	 * @param parent
+	 * @param sorted
+	 * @return returns only limited items.
+	 * @since 3.31
+	 */
+	protected Object[] applyItemsLimit(Object parent, Object[] sorted) {
+		// limit the number of items to be created. sorted always gets the remaining
+		// elements to be created.
+		int itemsLimit = getItemsLimit();
+		if (itemsLimit <= 0) {
+			return sorted;
+		}
+
+		if (sorted.length <= itemsLimit) {
+			return sorted;
+		}
+
+		int offSet = itemsLimit;
+		int srcPos = 0;
+
+		if (parent instanceof IExpandableNode expNode) {
+			srcPos = expNode.getOffSet();
+			offSet = srcPos + itemsLimit;
+			// pass on original children
+			sorted = expNode.getAllElements();
+		}
+
+		Object[] partialChildren = new Object[itemsLimit + 1];
+
+		// Extract a subset of children
+		System.arraycopy(sorted, srcPos, partialChildren, 0, itemsLimit);
+
+		// Add an expandable node
+		partialChildren[itemsLimit] = createExpandableNode(sorted, offSet, getItemsLimit());
+
+		return partialChildren;
+	}
+
+	/**
+	 * Viewer must dispose this cell and populate the remaining elements returned by
+	 * this cell.
+	 *
+	 * @param cell
+	 * @since 3.31
+	 */
+	protected void handleExpandableNodeClicked(Widget cell) {
+		// default implementation does nothing. Actual viewers can decide how to
+		// populate remaining elements.
+	}
+
+	/**
+	 * We should not dispose already visible items and we should display limited
+	 * elements along with already visible items.
+	 *
+	 * @param parent          model item to be refreshed.
+	 * @param visibleChildren currently visible children. This includes any elements
+	 *                        already expanded by user.
+	 * @return list of children to be displayed/refreshed.
+	 * @since 3.31
+	 */
+	protected Object[] getVisibleLimitBasedChildren(Object parent, Item[] visibleChildren) {
+		Object[] updatedChildren = null;
+		final int limit = getItemsLimit();
+		// at least one iteration of expandable items are populated.
+		if (visibleChildren.length > limit) {
+			List<Object> listOfEle = new ArrayList<>(Arrays.asList(getSortedChildren(parent)));
+			// case 1 . if returned are less then limit just stop here and add list of ele
+			// to updatedChildren
+			if (listOfEle.size() < limit) {
+				updatedChildren = new Object[listOfEle.size()];
+				listOfEle.toArray(updatedChildren);
+			} else {
+				// numberOfTimesRead is equal to number of times it was expanded by user.
+				int numberOfTimesRead = visibleChildren.length / limit;
+
+				if (visibleChildren.length % limit > 0
+						&& !(visibleChildren[visibleChildren.length - 1].getData() instanceof IExpandableNode)) {
+					numberOfTimesRead++;
+				}
+				List<Object> newList = new ArrayList<>();
+				Object ele1 = parent;
+
+				while (numberOfTimesRead > 0) {
+					List<Object> subList = new ArrayList<>(Arrays.asList(getSortedChildren(ele1)));
```

Comment:
getSortedChildren is called here second time with same (parent) argument. That should be avoided.